From: http://xenbits.xen.org/hg/linux-2.6.18-xen.hg/rev/cdbf3be3f2bd
# HG changeset patch
# User Olaf Hering <olaf@aepfle.de>
# Date 1326102808 -3600
# Node ID cdbf3be3f2bd1e3988ea3f1c795e5636c0b53d4b
# Parent  b4e1b9ff189f3b04475a1d0118932316abf26e19
Subject: xenbus: check availibility of XS_RESET_WATCHES command
Patch-mainline: n/a

Check platform-feature-xs_reset_watches before sending XS_RESET_WATCHES
command. Buggy xenstored implementations such as EC2 do not ignore unknown
commands properly and cause a guest hang.

Signed-off-by: Olaf Hering <olaf@aepfle.de>
Acked-by: jbeulich@suse.com

--- sle11sp2-2012-01-11.orig/drivers/xen/xenbus/xenbus_xs.c	2012-01-11 12:01:47.000000000 +0100
+++ sle11sp2-2012-01-11/drivers/xen/xenbus/xenbus_xs.c	2012-01-11 12:01:53.000000000 +0100
@@ -630,7 +630,13 @@ static struct xenbus_watch *find_watch(c
 static void xs_reset_watches(void)
 {
 #ifndef CONFIG_XEN
-	int err;
+	int err, supported = 0;
+
+	err = xenbus_scanf(XBT_NIL, "control",
+			   "platform-feature-xs_reset_watches", "%d",
+			   &supported);
+	if (err != 1 || !supported)
+		return;
 
 	err = xs_error(xs_single(XBT_NIL, XS_RESET_WATCHES, "", NULL));
 	if (err && err != -EEXIST)
